src/util-interval-tree.c

```diff
@@ -0,0 +1,421 @@
+{
+    for (uint32_t sig = 0; sig < sgh->init->max_sig_id + 1; sig++) {
+        if (sgh->init->sig_array[sig / 8] & (1 << (sig % 8))) {
+            printf(" %u", de_ctx->sig_array[sig]->id);
+        }
+    }
+}
+
+int COUNT = 4;
+void printIT(DetectEngineCtx *de_ctx, SCIntervalNode *node, int space,
+        void (*Print)(DetectEngineCtx *, SigGroupHead *));
+void printIT(DetectEngineCtx *de_ctx, SCIntervalNode *node, int space,
+        void (*Print)(DetectEngineCtx *, SigGroupHead *))
+{
+    // Base case
+    if (node == NULL)
+        return;
+
+    // Increase distance between levels
+    space += COUNT;
+
+    // Process right child first
+    printIT(de_ctx, IRB_RIGHT(node, irb), space, Print);
+
+    // Print current node after space
+    // count
+    printf("\n\n");
+    for (int i = COUNT; i < space; i++)
+        printf(" ");
+    printf("[%d, %d]: %s; p = %d; max = %d, sids:", node->port, node->port2,
+            IRB_COLOR(node, irb) ? "RED" : "BLACK",
+            IRB_PARENT(node, irb) ? IRB_PARENT(node, irb)->port : 0, node->max);
+    Print(de_ctx, node->sh);
+    printf("\n");
+
+    // Process left child
+    printIT(de_ctx, IRB_LEFT(node, irb), space, Print);
+}
+
+/**
+ * \brief Function to find all the overlaps of given ports with the existing
+ *        port ranges in the interval tree
+ * \param de_ctx Detection Engine Context
+ * \param port Given low port
+ * \param port2 Given high port
+ * \param ptr Pointer to the root of the tree
+ * \param list A list of DetectPort objects to be filled
+ */
+static void FindOverlaps(DetectEngineCtx *de_ctx, uint16_t port, uint16_t port2,
+        SCIntervalNode *root, DetectPort **list)
+{
+    SCLogDebug("\nfind overlaps for [%u:%u]", port, port2);
+    SCIntervalNode *ptr = root;
+    DetectPort *new_port = NULL;
+    int cnt_overlaps = 0;
+#if 0
+    printIT(de_ctx, root, 1, printIT_SigGroupHeadSigs);
+#endif
+    SCIntervalNode *stack[100], *current = root;
+    memset(&stack, 0, sizeof(stack));
+    int stack_depth = 0;
+    //            stack[stack_depth++] = current;
+
+    //    SCLogDebug("ROOT! [%u:%u]: current:%p", current->port, current->port2, current);
+    while (current || stack_depth) {
+        SCLogDebug("stack_depth %d", stack_depth);
+        for (int x = 0; x < stack_depth && x < 100 && stack[x] != NULL; x++)
+            SCLogDebug("stack[%d]: %p", x, stack[x]);
+
+        while (current != NULL) {
+            SCLogDebug("%s! [%u:%u]: current:%p",
+                    IRB_PARENT(current, irb) == NULL ? "ROOT" : "LEAF", current->port,
+                    current->port2, current);
+            if (current->max < port) {
+                current = NULL;
+                break;
+            }
+
+            ptr = current;
+            const bool is_overlap = IsOverlap(port, port2, ptr);
+            if (is_overlap) {
+                cnt_overlaps++;
+            }
+
+            if (is_overlap && (cnt_overlaps == 1)) {
```

Comment:
could this check for new_port == NULL instead?

Reply:
done.